godot-core/src/task/futures.rs

```diff
@@ -308,8 +335,162 @@ impl<C: WithBaseField, R: ParamTuple + Sync + Send> IntoFuture for &TypedSignal<
     }
 }
 
+/// Convert a value into a type that is [`Send`] at compile-time while the value might not be.
+///
+/// This allows to turn any implementor into a type that is `Send`, but requires to also implement [`DynamicSend`] as well.
+/// The later trait will verify if a value can actually be sent between threads at runtime.
+pub trait IntoMaybeSend: Sealed {
+    type Target: Send + DynamicSend<Target = Self>;
+
+    fn into_maybe_send(self) -> Self::Target;
+}
+
+/// Check that it was safe to send the value to the current thread.
+///
+/// # Safety
+/// The implementor has to guarantee that `assert_safe_send` panics if the value has been sent between threads while being `!Send`.
+pub unsafe trait DynamicSend: Send + Sealed {
+    type Target;
+
+    fn extract_if_safe(self) -> Option<Self::Target>;
+}
+
+pub struct ThreadConfined<T> {
+    value: T,
+    thread_id: ThreadId,
+}
+
+unsafe impl<T> Send for ThreadConfined<T> {}
+unsafe impl<T> Sync for ThreadConfined<T> {}
```

Comment:
Please add `// SAFETY: ` comment -- see also my previous comment about why we are allowed to add `Send` here.

Also, how does `Sync` play into all of this? Something being `Send` doesn't automatically make it `Sync` :thinking: 
(classic example: [`std::cell::Cell`](https://doc.rust-lang.org/std/cell/struct.Cell.html))

Reply:
yes, right, `Sync` shouldn't be here.